File: transfers/legit/routines/credit_cards.py
import transfers.credit_cards as credit_cards_transfer
from common.transactions import Transaction
from transfers.factory import TransactionFactory

from transfers.legit.blueprints import Blueprint, LegitBuildPlan


def generate_credit_lifecycle_txns(
    request: Blueprint,
    plan: LegitBuildPlan,
    txf: TransactionFactory,
    existing_txns: list[Transaction],
) -> list[Transaction]:
    cards = request.cc_state.cards
    if not request.cc_state.enabled() or cards is None:
        return []

    lifecycle = request.specs.cc_profile

    return credit_cards_transfer.generate(
        lifecycle.terms,
        lifecycle.habits,
        credit_cards_transfer.Request(
            window=request.timeline.window,
            rng=request.timeline.rng,
            base_seed=plan.seed,
            cards=cards,
            txns=existing_txns,
            primary_accounts=plan.primary_acct_for_person,
            issuer_acct=plan.counterparties.issuer_acct,
            txf=txf,
        ),
    )

File: src/transfers/legit/routines/credit_cards.cpp
#include "phantomledger/transfers/legit/routines/credit_cards.hpp"

#include "phantomledger/primitives/random/factory.hpp"
#include "phantomledger/transfers/channels/credit_cards/lifecycle.hpp"

#include <stdexcept>
#include <unordered_map>

namespace PhantomLedger::transfers::legit::routines::credit_cards {

namespace {

[[nodiscard]] time::Window
windowFromPlan(const blueprints::LegitBlueprint &plan) noexcept {
  return time::Window{
      .start = plan.startDate(),
      .days = plan.days(),
  };
}

[[nodiscard]] std::unordered_map<entity::PersonId, entity::Key>
primaryAccountKeysByPerson(const blueprints::LegitBlueprint &plan) {
  std::unordered_map<entity::PersonId, entity::Key> out;
  if (plan.allAccounts() == nullptr) {
    return out;
  }

  out.reserve(plan.primaryAcctRecordIx().size());
  for (const auto &kv : plan.primaryAcctRecordIx()) {
    const auto &record = plan.allAccounts()->records[kv.second];
    out.emplace(kv.first, record.id);
  }
  return out;
}

} // namespace

std::vector<transactions::Transaction>
generateLifecycle(const LifecycleRunRequest &request,
                  const blueprints::LegitBlueprint &plan,
                  const transactions::Factory &txf,
                  std::span<const transactions::Transaction> existingTxns) {
  if (request.cards == nullptr || request.cards->records.empty()) {
    return {};
  }
  if (request.rng == nullptr) {
    throw std::invalid_argument("credit_cards routine requires a non-null rng");
  }

  if (request.lifecycle == nullptr) {
    throw std::invalid_argument(
        "credit_cards routine requires lifecycle rules");
  }

  const auto primaryByPerson = primaryAccountKeysByPerson(plan);

  ::PhantomLedger::transfers::credit_cards::LedgerView view{
      .cards = *request.cards,
      .primaryAccounts = primaryByPerson,
      .issuerAccount = plan.counterparties().issuerAcct,
  };

  const random::RngFactory rngFactory{plan.seed()};

  ::PhantomLedger::transfers::credit_cards::Lifecycle lifecycle{
      *request.lifecycle,
      txf,
      rngFactory,
      view,
  };

  return lifecycle.generate(windowFromPlan(plan), existingTxns);
}

} // namespace PhantomLedger::transfers::legit::routines::credit_cards
